### DISCUSSION TITLE
rtl-mp

### DISCUSSION THREAD

**Discussion by bittnada** [The-OpenROAD-Project/OpenROAD#2445] [Q&A]
Hello,

I am trying to do floorplanning with OpenROAD with ISPD benchmark superblue16, but I am struggling with so many problems.
I am quite new to this filed, so my knowledge for this filed is not high enough.
I uploaded .lef, .def. and .v field given by ISPD and tried to do floorplan, but  this error line "LEF MASTER ~CELL HAS NO LIBERTY CELL" appeared and does not work at all. 
I am glad if anyone has any idea about this error and how to work with this kind of open bench mark.

Thanks

**Comment by maliberty** [comment]
I don't see any uploaded files.  Based on the message you have a cell called "~CELL" that exists in LEF but for which you have no .lib.

**Reply by rovinski** [reply]
Also keep in mind that you should not read both `.def` and `.v`. `.def` contains all the information from a `.v` netlist, so it's unnecessary.
Second, you did not mention a `.lib` file, which is required.

**Reply by bittnada** [reply]
Thanks for your kind reply. As you pointed out, I don't have .lib file since this kind of format was not given by ISPD contest. My project is aimed at running OpenROAD with open bench mark sample. I obtained a couple of open bench mark samples fro ISPD contest but all of them are lack of ".lib" file.
is .".lib" mandatory to run floorplanning in OpenROAD?   
I would appreciate if anyone can let me know where I can get open benchmark sample containing all of input file formats (.lef, .def, .v, .lib etc) for OpenROAD.

**Reply by bittnada** [reply]
> I don't see any uploaded files. Based on the message you have a cell called "~CELL" that exists in LEF but for which you have no .lib.

I tried OpenROAD with the files given by ISPD 2014 contest (Benchmark Design Suite B: https://www.ispd.cc/contests/14/web/downloads.html). But it doesn't contain ".lib" file. Is this a problem?

**Reply by QuantamHD** [reply]
So I had a look through that particular design contest output, and there's simply not enough there to do a meaningful place and route job.

OpenROAD has to have some concept of the timing between cells to do any kind of STA which is required for a number of steps in the flow.

The good news is that the following ISPD design contest in 2015 has the required files. http://iccad-contest.org/2015/problem_C/default.html
<img width="1409" alt="image" src="https://user-images.githubusercontent.com/3784748/199629862-721c5d7a-03ec-4ca9-b964-db866001eecc.png">

The order of commands required is as follows

```
read_def superblue16.def
read_liberty superblue16_Early.lib
read_liberty superblue16_Late.lib
read_sdc superblue16.sdc # This is optional/might not work

# Do whatever else you would like to do with the design
# As an example the command below will run timing driven placement
# global_placement -timing_driven -routability_driven -density 0.5 -pad_left 1-pad_right 1
```

**Reply by rovinski** [reply]
It might be possible to turn off global placement (GPL) features to allow it to run without a `.lib`, but the far better solution as @QuantamHD mentions above is to use the website above which seems to have a `.lib` to go along with the benchmark you mentioned (superblue16).

Without a `.lib`, you would not be able to get any further than global and detailed placement (GPL/DPO) because almost all subsequent steps require timing information from the `.lib`.

**Comment by bittnada** [comment]
Thanks to all of you
I followed the order of commands that @QuantamHD  recommended and it worked although LEF file seems to have some syntax errors.
Global placement (GPL) worked well, so I am very happy with this result. As @rovinski mentioned, global placement also worked without importing .lib file. Regarding with global placement, I have none of problem. I appreciate to all of you.

I am also trying to run floorplanning with openroad as I want to follow the paper (https://vlsicad.ucsd.edu/Publications/Conferences/389/c389.pdf), embedded in the floorplan machine for openroad.

Although the ISPD contest porblem is provided as "floorplan" finished, I want to re-floorplan with openroad and
compare the results between openroad and as it is given. 

I tried the commands as shown below,
```
read_verilog superblue16.v
read_lef tech.lef
read_lef superblue16.lef
read_lib superblue16_Early.lib
link_design superblue16
write_def superblue16.def
read_def superblue16.def -floorplan_initialize
partition_design -max_num_inst 1000 -min_num_inst 200 -max_num_macro 1 -min_num_macro 0 -net_threshold 0 -virtual_weight 500 -report_directory . -report_file report
```
Am I on the right track?
